text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,301 @@
+
+Most of the values of primitive Rust types have structural equality (integers, `bool`, `char`, references), but two families of types need special consideration:
+
+- Pointer types (raw pointers and function pointers): these compare by test the memory address for equality.
+  It is unclear whether that should be considered "structural", but it is fairly clear that this should be considered a bad idea:
+  Rust makes basically no guarantees for when two function pointers are equal or unequal
+  (the "same" function can be duplicated across codegen units and this have different addresses,
+  and different functions can be merged when they compile to the same assembly and thus have the same address).
+  Similarly, there are no or few guarantees for equality of pointers that are generated in constants.
+  However, there *is* a very clear notion of equality on pointers like `4 as *const i32`, and such pointers are occasionally used as sentinel value and used in `match`.
+  This is common enough to occur [even in the standard library](https://github.com/rust-lang/rust/blob/b18db7a13e52f71e94bdf221a7a013fd9ace4c7f/library/std/src/sys/windows/thread_parking.rs#L225-L226).
+  Thus we declare that values of raw pointer type have structural equality only if they are such pointers created by casting an integer.
+  Values of function pointer type never have structural equality;
+  it is very unusual to "cast" (really: transmute) an integer into a function pointer, and there is currently not sufficient motivation for allowing this in a pattern.
+- Floating-point types: in `f32` and `f64`, NaNs are not equal to themselves.
+  Furthermore, `+0.0` (written just `0.0` in Rust) and `-0.0` have different bit representations, but they *do* compare equal.
+  We can easily declare NaNs to not have structural equality, and reject them in patterns, since there is no situation where it makes sense to have a NaN in a pattern.
+  However, for zeroes this is more tricky -- allowing matching on `1.0` but rejecting `0.0` is likely going to be considered extremely arbitrary and inconsistent.
+  This RFC therefore suggests that all zeroes should have structural equality, and match all zeroes.
+  `0.0` and `-0.0` as a value of a constant used in a pattern will all match both `0.0` and `-0.0`, as they do today.
+  (As a literal, `0.0` and `-0.0` are currently both accepted but `+0.0` is not. This RFC does not propose to change anything here, though we could consider linting against `-0.0`.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When a constant `C` of type `T` is used as a pattern, we first check that `T: PartialEq`.
+Furthermore we require that the value of `C` *has (recursive) structural equality*, which is defined recursively as follows:
+
+- Integers, `bool` and `char` values, and references always have structural equality.
```

Comment:
I think this isn't quite correct: just because a value of type `&Foo` is a reference, doesn't mean it has structural equality. the pointed-to value also needs to have structural equality.

Reply:
*oops* you are right of course...